File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int good(TreeNode* root, int max) {
        if (root == nullptr) {
            return 0;
        }
        if (root->val >= max) {
            return 1 + good(root->left, root->val) + good(root->right, root->val);
        } else {
            return good(root->left, max) + good(root->right, max);
        }
    }
    int goodNodes(TreeNode* root) {
        int max = INT_MIN;
        return good(root, max);
    }
};

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int goodNodes(TreeNode root) {
        return solve(root, Integer.MIN_VALUE);
    }
    
    int solve(TreeNode root, int max) {
        if (root == null) {
            return 0;
        }
        if (root.val >= max) {
            return 1 + solve(root.left, root.val) + solve(root.right, root.val);
        }
        return solve(root.left, max) + solve(root.right, max);
    }
}